module_utils/network_lsr/argument_validator.py

```diff
@@ -1584,7 +1615,7 @@ def _validate_post(self, value, name, result):
                         "invalid 'interface_name' '%s'" % (result["interface_name"]),
                     )
             else:
-                if not result.get("mac"):
+                if not result.get("mac") and not result.get("match"):
```

Comment:
Should this also check whether anything in match is actually set? Would be good to have an example that uses an inventory with two machines and the same playbook. For one machine, a path match is specified in the inventory, for the other one there is not. Will this work as expected?

Reply:
Maybe we should validate `match` setting in argvalidator. When `match` is not `None`, at least one of the sub-setting (e.g. `path`) is not None, otherwise, raise an error.

Reply:
And I realize that if `match` is not actually set (e.g. `match` is `None`) or `path` is `None`, we will run into the following errors:

 `"fatal error: configuration error: connections[0].match: invalid content is not a dictionary"`
 https://github.com/linux-system-roles/network/blob/main/module_utils/network_lsr/argument_validator.py#L341
 
 `fatal error: 'NoneType' object is not iterable"`
 https://github.com/linux-system-roles/network/blob/main/module_utils/network_lsr/argument_validator.py#L386

Reply:
I was also planning to override ` _validate_post()` for `ArgValidator_DictMatch(ArgValidatorDict)`, but it seems that it was not necessary any more

```python
    def _validate_post(self, value, name, result):
        if not any(result[match_key] for match_key in result):
            raise ValidationError(
                name,
                "the value for match setting {0} can not be all set to 'None'".format(
                    list(result.keys())
                ),
            )

        return result
```

Reply:
I created two virtual machines via Virtual Box, one machine with pci address `pci-0000:00:03.0`, the other machine with   pci address `pci-0000:00:08.0`.  The test worked as excepted.

/etc/hosts
```
10.0.0.196 Fc33_template
10.0.0.108 Fc33_Clone1
```

Reply:
Thank you. Please test this by setting the PCI path only in Inventory.ini instead of the playbook and use a variable in the playbook.

Reply:
Inventory:
```
[all:vars]
pci_addresses=pci-0000:00:08.0

[VMs]
Fc33_template ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org
Fc33_Clone1 ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org
```

Test example:
```yaml
---
- hosts: VMs
  become: yes
  become_method: sudo
  vars:
    network_connections:
      - name: pciTest6
        type: ethernet
        autoconnect: yes
        state: up
        persistent_state: present
        match:
          path:
            - "{{ pci_addresses }}"
        ip:
          route_metric4: 100
          dns:
            - 10.0.0.3
          dns_search:
            - examples.com
            - subdomains.example.com
          dns_options:
            - rotate
            - timeout:2
  roles:
    - linux-system-roles.network
    ```

Test result:
    ```
TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
changed: [Fc33_Clone1] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.3"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["examples.com", "subdomains.example.com"], "route_metric4": 100}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107\n[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (is-modified)\n[005] <info>  #0, state:up persistent_state:present, 'pciTest6': connection reapplied\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107", "[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (is-modified)", "[005] <info>  #0, state:up persistent_state:present, 'pciTest6': connection reapplied"]}
[WARNING]: [003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020
[WARNING]: [004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)
[WARNING]: [005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this
connection (device lo not available because device is strictly unmanaged). (3)
fatal: [Fc33_template]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.3"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["examples.com", "subdomains.example.com"], "route_metric4": 100}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "msg": "error: up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)", "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020\n[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)\n[005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020", "[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)", "[005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)"]}
 
PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=8    changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0   
Fc33_template              : ok=5    changed=0    unreachable=0    failed=1    skipped=5    rescued=0    ignored=0  
```

Reply:
Inventory:
```
[all:vars]
pci_addresses=pci-0000:00:08.0

[VMs]
Fc33_template ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org
Fc33_Clone1 ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org
```

Test example:
```yaml
---
- hosts: VMs
  become: yes
  become_method: sudo
  vars:
    network_connections:
      - name: pciTest6
        type: ethernet
        autoconnect: yes
        state: up
        persistent_state: present
        match:
          path:
            - "{{ pci_addresses }}"
        ip:
          route_metric4: 100
          dns:
            - 10.0.0.3
          dns_search:
            - examples.com
            - subdomains.example.com
          dns_options:
            - rotate
            - timeout:2
  roles:
    - linux-system-roles.network
    ```

Test result:
    ```
TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
changed: [Fc33_Clone1] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.3"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["examples.com", "subdomains.example.com"], "route_metric4": 100}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107\n[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (is-modified)\n[005] <info>  #0, state:up persistent_state:present, 'pciTest6': connection reapplied\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107", "[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (is-modified)", "[005] <info>  #0, state:up persistent_state:present, 'pciTest6': connection reapplied"]}
[WARNING]: [003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020
[WARNING]: [004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)
[WARNING]: [005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this
connection (device lo not available because device is strictly unmanaged). (3)
fatal: [Fc33_template]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.3"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["examples.com", "subdomains.example.com"], "route_metric4": 100}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "msg": "error: up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)", "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020\n[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)\n[005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest6': add connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020", "[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 59af2e0f-c95a-4d48-b644-77f29f156020 (not-active)", "[005] <error> #0, state:up persistent_state:present, 'pciTest6': up connection failed: failure to activate connection: nm-manager-error-quark: No suitable device found for this connection (device lo not available because device is strictly unmanaged). (3)"]}
 
PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=8    changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0   
Fc33_template              : ok=5    changed=0    unreachable=0    failed=1    skipped=5    rescued=0    ignored=0  
```

Reply:
Thank you, this specifies the path for all machines. Please test it as I requested originally:  For one machine, a path match is specified in the inventory, for the other one there is not.

Reply:
The test worked as excepted.

Inventory:
```
[VMs]
Fc33_template ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org 
Fc33_Clone1 ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org pci_addresses=pci-0000:00:08.0
```

Test result:
```
TASK [linux-system-roles.network : Install packages] ***************************************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:20
fatal: [Fc33_template]: FAILED! => {"msg": "The conditional check 'not network_packages is subset(ansible_facts.packages.keys())' failed. The error was: error while evaluating conditional (not network_packages is subset(ansible_facts.packages.keys())): {{ __network_provider_setup[network_provider]['packages'] }}: {'nm': {'service_name': '{{ __network_service_name_default_nm }}', 'packages': '{{ __network_packages_default_nm }}'}, 'initscripts': {'service_name': '{{ __network_service_name_default_initscripts }}', 'packages': '{{ __network_packages_default_initscripts }}'}}: {{['NetworkManager'] + __network_packages_default_gobject_packages|select()|list() + __network_packages_default_wpa_supplicant|select()|list() + __network_packages_default_wireless|select()|list() + __network_packages_default_team|select()|list()}}: ['{% if __network_wpa_supplicant_required %}wpa_supplicant{% endif %}']: {{ __network_ieee802_1x_connections_defined or __network_wireless_connections_defined }}: {{ network_connections | selectattr('ieee802_1x', 'defined') | list | count > 0 }}: [{'name': 'pciTest6', 'type': 'ethernet', 'autoconnect': True, 'state': 'up', 'persistent_state': 'present', 'match': {'path': ['{{ pci_addresses }}']}, 'ip': {'route_metric4': 100, 'dns': ['10.0.0.4'], 'dns_search': ['exampl.com', 'subdo.example.com'], 'dns_options': ['rotate', 'timeout:2']}}]: 'pci_addresses' is undefined\n\nThe error appears to be in '/home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml': line 20, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n# Therefore install packages only when rpm does not find them\n- name: Install packages\n  ^ here\n"}
skipping: [Fc33_Clone1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
changed: [Fc33_Clone1] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.4"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["exampl.com", "subdo.example.com"], "route_metric4": 100}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107\n[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (not-active)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest6': update connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107", "[004] <info>  #0, state:up persistent_state:present, 'pciTest6': up connection pciTest6, 497b0a13-dbb5-490e-96bd-d67c24e29107 (not-active)"]}

PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=8    changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0   
Fc33_template              : ok=4    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0 
```

Reply:
It failed:

```
Fc33_template              : ok=4    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0 
```

Please adjust the playbook to make it properly configure the profile on the system without the pci address match for the template host.

Reply:
Test Example: (interface_name 'enp0s3' exists)
```yaml
- hosts: VMs
  become: yes
  become_method: sudo
  tasks:
    - name: "Template host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: enp0s3
            type: ethernet
            autoconnect: yes
            state: up
            persistent_state: present
            ip:
              route_metric4: 100
              dns:
                - 10.0.0.5
              dns_search:
                - exampl.com
                - subdo.example.com
              dns_options:
                - rotate
                - timeout:2
      when: "'template' in inventory_hostname"

    - name: "Clone host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: enp0s3
            type: ethernet
            autoconnect: yes
            state: up
            persistent_state: present
            match:
              path:
                - "{{ pci_addresses }}"
            ip:
              route_metric4: 100
              dns:
                - 10.0.0.5
              dns_search:
                - exampl.com
                - subdo.example.com
              dns_options:
                - rotate
                - timeout:2
      when: "'Clone' in inventory_hostname"
```

Test result:  (interface_name 'enp0s3' exists)
```
PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=8    changed=1    unreachable=0    failed=0    skipped=17   rescued=0    ignored=0   
Fc33_template              : ok=8    changed=1    unreachable=0    failed=0    skipped=17   rescued=0    ignored=0   
```

Test Example: (interface name 'pciTest6' not exists)
 ```yaml
 - hosts: VMs
  become: yes
  become_method: sudo
  tasks:
    - name: "Template host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: pciTest6
            type: ethernet
            autoconnect: yes
            state: up
            persistent_state: present
            ip:
              route_metric4: 100
              dns:
                - 10.0.0.5
              dns_search:
                - exampl.com
                - subdo.example.com
              dns_options:
                - rotate
                - timeout:2
      when: "'template' in inventory_hostname"

    - name: "Clone host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: pciTest6
            type: ethernet
            autoconnect: yes
            state: up
            persistent_state: present
            match:
              path:
                - "{{ pci_addresses }}"
            ip:
              route_metric4: 100
              dns:
                - 10.0.0.5
              dns_search:
                - exampl.com
                - subdo.example.com
              dns_options:
                - rotate
                - timeout:2
      when: "'Clone' in inventory_hostname"
```

Test result: (interface name 'pciTest6' not exists)
```
TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
skipping: [Fc33_Clone1] => {"changed": false, "skip_reason": "Conditional result was False"}
[WARNING]: [001] <error> #0, state:up persistent_state:present, 'pciTest6': profile specifies interface_name 'pciTest6' but no such interface exists
fatal: [Fc33_template]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": true, "ip": {"dns": ["10.0.0.5"], "dns_options": ["rotate", "timeout:2"], "dns_search": ["exampl.com", "subdo.example.com"], "route_metric4": 100}, "name": "pciTest6", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": false, "msg": "error: profile specifies interface_name 'pciTest6' but no such interface exists", "stderr": "[001] <error> #0, state:up persistent_state:present, 'pciTest6': profile specifies interface_name 'pciTest6' but no such interface exists\n", "stderr_lines": ["[001] <error> #0, state:up persistent_state:present, 'pciTest6': profile specifies interface_name 'pciTest6' but no such interface exists"]}

PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=8    changed=1    unreachable=0    failed=0    skipped=17   rescued=0    ignored=0   
Fc33_template              : ok=5    changed=0    unreachable=0    failed=1    skipped=5    rescued=0    ignored=0   
```

Reply:
Thank you. Please let me clarify: There should be only one profile and on one machine it should match according to the pci address, on the other machine without the pci address but it should still work. This should be controlled by the inventory and the playbook should be as simple as possible. It should not have specific profiles for each machine but a generic one that can be used for all machines depending on the inventory variables (here the variables only need to specify PCI address or maybe interface name (if they are different)).

Reply:
This will work, but I am afraid that this is also considered to be two connection profiles, right ? 
```yaml
- hosts: VMs
  become: yes
  become_method: sudo
  tasks:
    - name: "host configure with PCI"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: enp0s3
            type: ethernet
            state: up
            match: {path: ["{{ pci_addresses }}"]}
            ip:
              route_metric4: 10
      when: pci_addresses is defined

    - name: "host configure without PCI"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: enp0s3
            type: ethernet
            state: up
            ip:
              route_metric4: 10
      when: pci_addresses is undefined
```

Reply:
But if we clearly define the `match` setting to be `None`, the tests failed:

```yaml
        network_connections:
          - name: enp0s3
            type: ethernet
            state: up
            match: null
            ip:
              route_metric4: 10
```

Result:
```
TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
fatal: [Fc33_Clone1]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": null, "name": "enp0s3", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": false, "msg": "fatal error: configuration error: connections[0].match: invalid content is not a dictionary", "stderr": "\n", "stderr_lines": [""]}
fatal: [Fc33_template]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": null, "name": "enp0s3", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": false, "msg": "fatal error: configuration error: connections[0].match: invalid content is not a dictionary", "stderr": "\n", "stderr_lines": [""]}

PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=5    changed=0    unreachable=0    failed=1    skipped=5    rescued=0    ignored=0   
Fc33_template              : ok=5    changed=0    unreachable=0    failed=1    skipped=5    rescued=0    ignored=0 
```

Reply:
Finally, it works. 

Test example:
```yaml
- hosts: VMs
  become: yes
  become_method: sudo
  tasks:
    - name: "Assign match.path value when pci_addresses is defined"
      set_fact:
        pci_addresses: "{{ [pci_addresses] }}"
      when: pci_addresses is defined

    - name: "Assign match.path value when pci_addresses is not defined"
      set_fact:
        pci_addresses: "{{ [] }}"
      when: pci_addresses is not defined

    - name: "host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: pciTest8
            type: ethernet
            state: up
            match:
              path: "{{ pci_addresses }}"
            ip:
              route_metric4: 10
```              
Test result:
```
TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
changed: [Fc33_template] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": {"path": []}, "name": "pciTest8", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest8': add connection pciTest8, ab20b9f9-5ed8-4108-9f91-0ec973a94a2b\n[004] <info>  #0, state:up persistent_state:present, 'pciTest8': up connection pciTest8, ab20b9f9-5ed8-4108-9f91-0ec973a94a2b (not-active)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest8': add connection pciTest8, ab20b9f9-5ed8-4108-9f91-0ec973a94a2b", "[004] <info>  #0, state:up persistent_state:present, 'pciTest8': up connection pciTest8, ab20b9f9-5ed8-4108-9f91-0ec973a94a2b (not-active)"]}
changed: [Fc33_Clone1] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest8", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest8': add connection pciTest8, be49b699-f179-4d1f-a95e-9b8d65b42583\n[004] <info>  #0, state:up persistent_state:present, 'pciTest8': up connection pciTest8, be49b699-f179-4d1f-a95e-9b8d65b42583 (not-active)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest8': add connection pciTest8, be49b699-f179-4d1f-a95e-9b8d65b42583", "[004] <info>  #0, state:up persistent_state:present, 'pciTest8': up connection pciTest8, be49b699-f179-4d1f-a95e-9b8d65b42583 (not-active)"]}

PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=9    changed=1    unreachable=0    failed=0    skipped=6    rescued=0    ignored=0   
Fc33_template              : ok=9    changed=1    unreachable=0    failed=0    skipped=6    rescued=0    ignored=0
```

Reply:
This might be more elegant test playbook:
```yaml
- hosts: VMs
  become: yes
  become_method: sudo
  tasks:
    - name: "host configure"
      import_role:
        name: linux-system-roles.network
      vars:
        network_connections:
          - name: pciTest10
            type: ethernet
            # autoconnect: yes
            persistent_state: present
            state: up
            match:
              path: "{{ [] if pci_addresses is not defined else [pci_addresses] }}"
            ip:
              route_metric4: 10
 ```
 
 Test result:
 ```
 TASK [linux-system-roles.network : Configure networking connection profiles] ***************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
changed: [Fc33_Clone1] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": {"path": ["pci-0000:00:08.0"]}, "name": "pciTest11", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest11': add connection pciTest11, 4f67bd2d-4a35-457a-bc02-c0eb4c59e13d\n[004] <info>  #0, state:up persistent_state:present, 'pciTest11': up connection pciTest11, 4f67bd2d-4a35-457a-bc02-c0eb4c59e13d (not-active)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest11': add connection pciTest11, 4f67bd2d-4a35-457a-bc02-c0eb4c59e13d", "[004] <info>  #0, state:up persistent_state:present, 'pciTest11': up connection pciTest11, 4f67bd2d-4a35-457a-bc02-c0eb4c59e13d (not-active)"]}
changed: [Fc33_template] => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"ip": {"route_metric4": 10}, "match": {"path": []}, "name": "pciTest11", "persistent_state": "present", "state": "up", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": true, "stderr": "[003] <info>  #0, state:up persistent_state:present, 'pciTest11': add connection pciTest11, ddb3dc68-3e19-408c-84c0-8f03b9f05af9\n[004] <info>  #0, state:up persistent_state:present, 'pciTest11': up connection pciTest11, ddb3dc68-3e19-408c-84c0-8f03b9f05af9 (not-active)\n", "stderr_lines": ["[003] <info>  #0, state:up persistent_state:present, 'pciTest11': add connection pciTest11, ddb3dc68-3e19-408c-84c0-8f03b9f05af9", "[004] <info>  #0, state:up persistent_state:present, 'pciTest11': up connection pciTest11, ddb3dc68-3e19-408c-84c0-8f03b9f05af9 (not-active)"]}

PLAY RECAP *********************************************************************************************************************************************************************************
Fc33_Clone1                : ok=9    changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0   
Fc33_template              : ok=9    changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0 
```

Reply:
After I dug into the code, I noticed:
https://github.com/linux-system-roles/network/blob/main/module_utils/network_lsr/argument_validator.py#L385

So

```
match:
  path: "{{ pci_address | default("") }}
```

should work. I believe I asked about using `path: "{{ pcie_address }}"` before but for some reason it is hard to find conversations/comments again in this PR.

Reply:
yes, default filer will work if we define the PCI address in inventory as list:
 ```
 [VMs]
Fc33_template ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org 
Fc33_Clone1 ansible_connection=ssh ansible_user=osboxes ansible_ssh_pass=osboxes.org ansible_sudo_pass=osboxes.org pci_address='["pci-0000:00:08.0"]'
```
And in the test playbook, we can write the match section as this way:
```yaml
            match:
              path: "{{ pci_address | default([]) }}"
```